### PR TITLE
Fix loading issue for Faster RCNN model from #16783

### PR DESCRIPTION
**Pull request by sl-sergei**
relates: #16783
**merge with extra:** opencv/opencv_extra#804

Necessary changes for further work on supporting RCNN-like models from **torchvision**.

<cut/>

```
opencv_extra=multi_output_gather
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2021.1.0:20.04
build_image:Custom Win=openvino-2021.1.0
build_image:Custom Mac=openvino-2021.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by sl-sergei** [comment]
@dkurt Please, take a look at the fix (seems like a crutch, but anyway). Still, it doesn't fix the problem on whole, because there are many unsupported operations in the ONNX model:
**Control Flow**
image
**Eltwise**
image
They cannot be reduced to the current implementations of Detection output layers, so I think it is a good idea to implement these layers as was suggested in https://github.com/opencv/opencv/pull/16750.
Can we discuss it somewhere outside Github?

**Comment by dkurt** [comment]
Even all the layers will be imported, model execution will be inefficient because of too atomic operations.

**Comment by sl-sergei** [comment]
> Even all the layers will be imported, model execution will be inefficient because of too atomic operations.

Yes, I understand it, but to prevent it I plan to implement aggregate layers for RCNN models in a way similar to current **Softmax** or **BatchNormalization** OpenCV layers. And even in this case extra layers are required.

**Comment by sl-sergei** [comment]
@mshabunin I think there is a problem with Linux Custom builder, there might be some problems with MYRIAD device, could you check?

**Comment by mshabunin** [comment]
@sl-sergei , there is a merge conflict in _opencv_extra_ repository now, please rebase your branch.

File: modules/dnn/src/graph_simplifier.cpp [hunk 1]
```diff
@@ -112,6 +112,23 @@ bool Subgraph::match(const Ptr<ImportGraphWrapper>& net, int nodeId,
                 continue;
             nodeId = getInputNodeId(net, node, j);
             const Ptr<ImportNodeWrapper> inpNode = net->getNode(nodeId);
+            if ((inpNode->getType() == "Gather") && (node->getType() == "Cast"))
```
Comment (dkurt) [inline modules/dnn/src/graph_simplifier.cpp:115]: Why we need to change generic part of graphs pattern match?

Comment (sl-sergei) [reply modules/dnn/src/graph_simplifier.cpp:115]: The pattern match is not correct when there are several Cast nodes that use Gather node as input. Right now, it merges Gather with one of the Cast nodes, and others no longer have a correct input node, so it causes an error.

Comment (dkurt) [reply modules/dnn/src/graph_simplifier.cpp:115]: The thing is that this is a common graphs fusion part but "Gather" and "Cast" are too specific.

Comment (sl-sergei) [reply modules/dnn/src/graph_simplifier.cpp:115]: This logic doesn't work specifically for the combination of "Gather" and "Cast" nodes. If you think that there is a better way for avoiding this error, please, share it, I will update the fix.

Comment (dkurt) [reply modules/dnn/src/graph_simplifier.cpp:115]: ONNX related transformations should be located at https://github.com/opencv/opencv/blob/master/modules/dnn/src/onnx/onnx_graph_simplifier.cpp

Comment (alalek) [reply modules/dnn/src/graph_simplifier.cpp:115]: @sl-sergei `graph_simplifier.cpp` is a generic template. **Specialization** for specific cases should go to another place.

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 2]
```diff
@@ -260,6 +260,38 @@ class GatherCastSubgraph : public Subgraph
...
     }
+
+    virtual bool match(const Ptr<ImportGraphWrapper>& net, int nodeId,
+                       std::vector<int>& matchedNodesIds,
+                       std::vector<int>& targetNodesIds) CV_OVERRIDE
+    {
+        bool retVal = Subgraph::match(net, nodeId, matchedNodesIds, targetNodesIds);
+        size_t matchedNodesNum = matchedNodesIds.size();
+        // Now we check if merging can be made for these Gather and Cast nodes
+        if (retVal && matchedNodesNum >= 2) {
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:271]: if (!retVal || matchedNodesNum < 2)
    return false;
// other logic

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 3]
```diff
@@ -260,6 +260,38 @@ class GatherCastSubgraph : public Subgraph
...
+            int nodeToMatch = matchedNodesIds[matchedNodesNum - 1];
+            const Ptr<ImportNodeWrapper> node = net->getNode(nodeToMatch);
+            if (node->getType() == "Cast") {
+                int inpNodeId = matchedNodesIds[matchedNodesNum - 2];
+                const Ptr<ImportNodeWrapper> inpNode = net->getNode(inpNodeId);
+                if ((inpNode->getType() == "Gather")) {
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:277]: extra `( )`

**Review by dkurt — APPROVED** [review]
:+1: Thank you!

**Comment by alalek** [comment]
Looks like this patch brokes MYRIAD tests somehow:
- good: http://pullrequest.opencv.org/buildbot/builders/3_4_openvino-skl-lin64/builds/857 ( 564d1a0f79b3a7a528138b990837290f6cacfdd6 )
- bad: http://pullrequest.opencv.org/buildbot/builders/3_4_openvino-skl-lin64/builds/856 ( 2b82f8f12c2ac6f66872e0e6316f8f0b21b6ee13 )
